hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/lock/BaseZookeeperBasedLockProvider.java

```diff
@@ -0,0 +1,196 @@
+ * using zookeeper. Users need to have a Zookeeper cluster deployed to be able to use this lock.
+ */
+@NotThreadSafe
+public abstract class BaseZookeeperBasedLockProvider implements LockProvider<InterProcessMutex>, Serializable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(BaseZookeeperBasedLockProvider.class);
+
+  private final transient CuratorFramework curatorFrameworkClient;
+  private volatile InterProcessMutex lock = null;
+  protected final LockConfiguration lockConfiguration;
+  protected final String zkBasePath;
+  protected final String lockKey;
+
+  public static final int MAX_ZK_BASE_PATH_NUM_BYTES = 4096;
+
+  public BaseZookeeperBasedLockProvider(final LockConfiguration lockConfiguration, final StorageConfiguration<?> conf) {
+    checkRequiredProps(lockConfiguration);
+    this.lockConfiguration = lockConfiguration;
+    zkBasePath = getZkBasePath(lockConfiguration);
+    lockKey = getLockKey(lockConfiguration);
+    this.curatorFrameworkClient = CuratorFrameworkFactory.builder()
+        .connectString(lockConfiguration.getConfig().getString(ZK_CONNECT_URL_PROP_KEY))
+        .retryPolicy(new BoundedExponentialBackoffRetry(lockConfiguration.getConfig().getInteger(LOCK_ACQUIRE_RETRY_WAIT_TIME_IN_MILLIS_PROP_KEY),
+            lockConfiguration.getConfig().getInteger(LOCK_ACQUIRE_RETRY_MAX_WAIT_TIME_IN_MILLIS_PROP_KEY),
+            lockConfiguration.getConfig().getInteger(LOCK_ACQUIRE_NUM_RETRIES_PROP_KEY)))
+        .sessionTimeoutMs(lockConfiguration.getConfig().getInteger(ZK_SESSION_TIMEOUT_MS_PROP_KEY, DEFAULT_ZK_SESSION_TIMEOUT_MS))
+        .connectionTimeoutMs(lockConfiguration.getConfig().getInteger(ZK_CONNECTION_TIMEOUT_MS_PROP_KEY, DEFAULT_ZK_CONNECTION_TIMEOUT_MS))
+        .build();
+    this.curatorFrameworkClient.start();
+    createPathIfNotExists();
+  }
+
+  protected abstract String getZkBasePath(LockConfiguration lockConfiguration);
+
+  protected abstract String getLockKey(LockConfiguration lockConfiguration);
+
+  protected String generateLogSuffixString() {
+    return StringUtils.join("ZkBasePath = ", zkBasePath, ", lock key = ", lockKey);
+  }
+
+  protected String getLockPath() {
+    return zkBasePath + '/' + lockKey;
+  }
+
+  private void createPathIfNotExists() {
+    try {
+      String lockPath = getLockPath();
+      LOG.info(String.format("Creating zookeeper path %s if not exists", lockPath));
+      String[] parts = lockPath.split("/");
+      StringBuilder currentPath = new StringBuilder();
+      for (String part : parts) {
+        if (!part.isEmpty()) {
+          currentPath.append("/").append(part);
+          createNodeIfNotExists(currentPath.toString());
+        }
+      }
+    } catch (Exception e) {
+      LOG.error("Failed to create ZooKeeper path: " + e.getMessage());
+      throw new HoodieLockException("Failed to initialize ZooKeeper path", e);
+    }
+  }
+
+  private void createNodeIfNotExists(String path) throws Exception {
```

Comment:
Do we want to keep the same logic as before:
```
private void createNodeIfNotExists(String path) throws Exception {
    if (this.curatorFrameworkClient.checkExists().forPath(path) == null) {
      try {
        this.curatorFrameworkClient.create().forPath(path);
        // to avoid failure due to synchronous calls.
      } catch (KeeperException e) {
        if (e.code() == KeeperException.Code.NODEEXISTS) {
          LOG.debug(String.format("Node already exist for path = %s", path));
        } else {
          throw new HoodieLockException("Failed to create zookeeper node", e);
        }
      }
    }
  }
```